jurigged/codedb.py

```diff
@@ -0,0 +1,156 @@
+import gc
+import sys
+import time
+import types
+
+from ovld import ovld
+
+##########
+# CodeDB #
+##########
+
+MAX_TIME = 0.1
+
+
+def make_audithook(self):
+    def watch_exec(event, obj):  # pragma: no cover
+        # Note: Python does not trace audit hooks, so normal use will not show
+        # coverage of this function even if it is executed
+        if event == "exec":
+            (code,) = obj
+            self.assimilate(code)
+
+    # The closure is faster than a method on _CodeDB
+    return watch_exec
+
+
+class _CodeDB:
+    def __init__(self):
+        self.codes = {}
+        self.functions = {}
+        self.last_cost = 0
+        self.always_use_cache = False
+
+    def setup(self):
+        self.collect_all()
+        sys.addaudithook(make_audithook(self))
+
+    def collect_all(self):
+        results = []
+        for obj in gc.get_objects():
+            if isinstance(obj, types.FunctionType):
+                results.append((obj, obj.__code__))
+            elif not hasattr(obj, "__getattr__"):
+                if hasattr(obj, "__conform__"):
+                    for x in gc.get_referents(obj):
+                        if isinstance(x, types.CodeType):
+                            results.append((obj, x))
+        for obj, co in results:
+            if hasattr(obj, "__qualname__"):
+                self.assimilate(
+                    co, (co.co_filename, *obj.__qualname__.split("."))
```

Comment:
I may be misunderstanding the code but this strikes me as off for several reasons:

1. `__qualname__` is often replaced, especially by functools.wraps. One way to see the effect of this is with this code:

```python
import jurigged

groups = {}
for path, code in jurigged.codedb.db.codes.items():
    groups.setdefault(code, set()).add(path)

for code, group in groups.items():
    if len(group) > 1:
        print(code)
        for path in group:
            print(path)
```

   This shows codes which are reached by many paths. You can see many such cases are wrappers returned by decorators.

2. The code `path = (*path, name)` under `assimilate` means that the name will be duplicated since it's also at the end of the qualname. For example the only path to `_CodeDB.setup` is `('jurigged/codedb.py', '_CodeDB', 'setup', 'setup', 34)`
3. The qualname includes `<locals>` which I don't think you want.

Reply:
Thinking about this some more, when you loop through `co_consts` and call `assimilate` recursively, that will cover most cases. Extra paths created by decorators with functools.wraps and qualnames containing `<locals>` create clutter in `.codes` but the correct paths should also be there and things should work. The real problems are:

1. Functions which are neither global nor defined within another function, i.e. they're defined in a class which itself is not in a function, which is most methods. Hence the `_CodeDB.setup` example above. It's never reached by the recursion in `assimilate`. Obviously this needs to be fixed but it also shows that you need test cases which cover module-level classes and their methods.
2. Theoretically, someone might replace `__qualname__` on the actual original function and not just a wrapper around the function. To guarantee the correct qualname you'd need to parse the source, infer the qualnames of definitions, and look up definitions by name and line number, much like the code you already have. That's what I do: https://github.com/alexmojaki/executing/blob/master/executing/executing.py#L389

Reply:
> The code path = (*path, name) under assimilate means that the name will be duplicated since it's also at the end of the qualname.

Oh, that's true, I should cut out the last element of qualname.

> Functions which are neither global nor defined within another function, i.e. they're defined in a class which itself is not in a function, which is most methods. Hence the _CodeDB.setup example above. It's never reached by the recursion in assimilate.

For the initial pass, gc.get_objects should contain the methods already. Other objects are acquired through the audit hook when modules are exec-ed, and methods *are* reaachable through that because classes also have code objects.

> Obviously this needs to be fixed but it also shows that you need test cases which cover module-level classes and their methods.

I do, e.g. `test_codetools::test_merge`. This being said, all of these are acquired through the audithook, which properly sets the path because it does not use qualname, so there was a blind spot there.

Reply:
I don't really understand what you're saying or the full details of how the code works, but it makes sense to me that in the end you'd look things up by qualname rather than line number since qualnames are more stable. If a file contains two functions with the same name and after an edit neither of those functions are at their original line, the containing class is a good way to distinguish. I'm talking about using the line numbers early on to determine the qualnames, before the source code is edited. I suppose I don't actually know when this code gets called so maybe that's nonsense.

Reply:
The qualname isn't stable for me either since I use `ovld` very liberally, so I have a whole lot of functions with the same qualnames.

I think I'm basically already doing what you're saying, though, it's just that I use both qualname and line number for the initial lookup because of issues about the moment the initial lookup is performed. With the audit hook, the "qualname" is inferred from the code objects, it's just for that initial pass with gc.get_objects that I have to get it from the function because I don't have the parent code objects.

The entry in codedb is looked up and cached by `FunctionDefinition` in codetools.py whenever it has to be merged with an updated `FunctionDefinition`. Matching up definitions is done using the name hierarchy found in the source and crude code similarity (I don't think the algorithm is very good, but it'll do for now), neither the Python qualname nor the line number really enter into the equation outside of the initial lookup.